blog/tidy-collaborative-filtering/index.qmd

```diff
@@ -0,0 +1,214 @@
+
+# Train Collaborative Filtering Recommenders in R
+
+For this post, we rely on the `tidyverse` [@Wickham2019] family of packages, `scales` [@scales] for scale functions for visualization, and `recommenderlab` [@recommenderlab] - a package that provides an infrastructure to develop and test collaborative filerting recommender algorithms.  
+```{r, message = FALSE, warning = FALSE}
+library(tidyverse)
+library(scales)
+library(recommenderlab)
+```
+
+```{r, include = FALSE}
+wikifolio_portfolios <- read_rds("../data/wikifolio_portfolios.rds")
+```
+
+I load a dataset with stock holdings of investable wikifolios at the beginning of 2023. Reach out if you want to get access to the data. 
+
+```{r}
+glimpse(wikifolio_portfolios)
+```
+
+First, I convert the long data to a binary rating matrix where only non-NA values are stored explicitly and NA values are represented by dots for efficiency. 
+
+```{r}
+binary_rating_matrix <- wikifolio_portfolios |>
+  pivot_wider(id_cols = wikifolio,
+              names_from = stock,
+              values_from = in_portfolio,
+              values_fill = list(in_portfolio = 0)) |>
+  select(-wikifolio) |>
+  as.matrix() |>
+  as("binaryRatingMatrix")
+binary_rating_matrix
+```
+
+As in our book chapter on factor selection via machine learning, I perform cross-validation and split the data into training and test data. The training sample constitute 80% of the data and I perform $k$-times recursive estimation of the cross validation with $k=5$. Note that the parameter `given=-1` means that all but one randomly selected item is ignored for the evaluation.
+
+```{r}
+scheme <- binary_rating_matrix |>
+  evaluationScheme(
+    method = "cross",
+    k      = 5,
+    train  = 0.8,
+    given  = -1
+)
+scheme
+```
+
+Here is the list of recommenders that I consider for the backtest:
+
+* Random Items: the benchmark case because it just stupidly chooses random stocks from all possible choices.
+* Popular Items: just recommended the most popular stocks to measured by the number of wikifolios that hold the stock.
+* Association Rules: each wikifolio and its portfolio is considered as a transaction. Association rule mining finds similar portfolios across all traders (if a traders has x, y and z in his/her portfolio, then he/she is X% likely of also including w). 
+* Item-Based Filtering: the algorithm calculates a similarity matrix across stocks. Recommendations are then based on the list of most similar stocks to the ones the wikifolio already has in its portfolio. 
+* User-Based Filtering: the algorithm finds a neighborhood of similar wikifolios for each wikifolio (for this exercise it is set to 100 most similar wikifolios). Recommendations are then based on what the most similar wikifolios have in their portfolio. 
+
+For each algorithm, I base the evaluation on 1, 3, 5, and 10 recommendations. Note that the evaluation takes a couple of hours, in particular because IBCF and UBCF are quite time-consuming. 
```

Comment:
Item-Based and User-Based Filtering are quite time-consuming - no need for abbreviations.